File: exercises/intro_to_ddd/java/src/main/java/domain/Motherboard.java
package domain;

import java.util.List;

public final class Motherboard implements Component {
    private final List<MemorySlot> memorySlots;

    public Motherboard(List<MemorySlot> memorySlots) {
        this.memorySlots = memorySlots;
    }

    public CompatibilityResult attachMemory(MemoryStick memory) {
        // Try to attach the memory stick to each memory slot
        for (MemorySlot slot : memorySlots) {

            // If attaching to a slot is successful, return a compatible result
            CompatibilityResult result = slot.attach(memory);
            if (result.isCompatible()) {
                return result;
            }
        }

        // If no slots were compatible, return an incompatible result
        return CompatibilityResult.incompatible(memory, "No available memory slots");
    }

    @Override
    public String toString() {
        return "Motherboard(" + memorySlots.size() + " slots)";
    }
}

File: exercises/intro_to_ddd/typescript/domain/Motherboard.ts
import { CompatibilityResult } from "./CompatibilityResult"
import type { Component } from "./Component"
import type { MemorySlot } from "./MemorySlot"
import type { MemoryStick } from "./MemoryStick"

export class Motherboard implements Component {
  memorySlots: MemorySlot[]

  constructor(memorySlots: MemorySlot[]) {
    this.memorySlots = memorySlots
  }

  attachMemory(memory: MemoryStick): CompatibilityResult {
    // Try to attach the memory stick to each memory slot
    for (const slot of this.memorySlots) {

      // If attaching to a slot is successful, return a compatible result
      if (slot.attach(memory).isCompatible()) {
        return CompatibilityResult.compatible(memory)
      }
    }

    // If no slots were compatible, return an incompatible result
    return CompatibilityResult.incompatible(memory, 'No available memory slots')
  }
}
